### ISSUE TITLE
stack overflow protection limit the max recursion depth when using custom frame evaluation

### ISSUE THREAD

**Issue by art049** [python/cpython#105003]
# Bug report

Enabling the perf trampoline messes with the recursion limit.

## Reproduction

```python
def recurse(n):
    if n == 0:
        return
    recurse(n - 1)


if __name__ == "__main__":
    import sys

    n = 801
    sys.setrecursionlimit(n)
    recurse(n - 2)
```

The following execution works perfectly:
```sh
python recurse.py
```

However, this gives a recursion error:
```sh
python -X perf recurse.py
```
```
Traceback (most recent call last):
  File "<...>/recurse.py", line 12, in <module>
    recurse(n - 2)
  File "<...>/recurse.py", line 4, in recurse
    recurse(n - 1)
  File "<...>/recurse.py", line 4, in recurse
    recurse(n - 1)
  File "<...>/recurse.py", line 4, in recurse
    recurse(n - 1)
  [Previous line repeated 796 more times]
RecursionError: maximum recursion depth exceeded
```

Strangely, the issue only arises when `n >= 801` and not under this value.

@pablogsal if you have an idea of what could create the problem, I'd be happy to work on a fix!

# Your environment

- CPython versions tested on: 3.12.0b1, 3.12.0-dev(2c02c68)
- OS: linux 5.15

**Comment by art049** [comment]
The problem seems to come from the C recursion limit added in https://github.com/python/cpython/issues/91079

This explains why the behavior happens only after a depth of 800:
https://github.com/python/cpython/blob/2c02c6886739f0ed420d900b2a29933bc1c5df37/Include/cpython/pystate.h#L258

The custom frame eval function set by the perf trampoline decreases `tstate->c_recursion_remaining`, changing the max recursion depth.
I'm not sure yet if it's specifically related to the `py_trampoline_evaluator` or if the problem appears more globally when a custom frame eval is used.

**Comment by sunmy2019** [comment]
Confirmed it's restricted by the C recursion limit.

I think the correct thing here is to expose an API to change this. But this sound a little more, and may not be backported to 3.12? @markshannon @gpshead 

```c
    int py_recursion_remaining;
    int py_recursion_limit;

    int c_recursion_remaining;
+    int c_recursion_limit;
```

**Comment by art049** [comment]
> I think the correct thing here is to expose an API to change this. But this sound a little more, and may not be backported to 3.12?

It might still break some existing codebases that were using a custom eval function, limiting their recursion depth to 800  without AFAIK any possible workaround.

**Comment by pablogsal** [comment]
> It might still break some existing codebases that were using a custom eval function, limiting their recursion depth to 800 without AFAIK any possible workaround.

Using a custom eval function and `perf` is not supported because `perf` support relies on a custom eval function. This means that unless you wrap it you are out of luck.

**Comment by art049** [comment]
> Using a custom eval function and `perf` is not supported because `perf` support relies on a custom eval function. This means that unless you wrap it you are out of luck.

Yes totally. I wasn't very clear but I also managed to reproduce the issue with another dummy custom eval function like this:
```c++
static PyObject * dummy_custom_evaluator(PyThreadState *ts, _PyInterpreterFrame *frame, int throw)
{
    return _PyEval_EvalFrameDefault(ts, frame, throw);
}
```

Thus for existing uses of the custom eval function, it might break existing implementations.
Since this issue is also affecting, the `perf` support, I'm not sure if I should create another issue.

**Comment by pablogsal** [comment]
> > Using a custom eval function and `perf` is not supported because `perf` support relies on a custom eval function. This means that unless you wrap it you are out of luck.
> 
> Yes totally. I wasn't very clear but I also managed to reproduce the issue with another dummy custom eval function like this:
> 
> ```c++
> static PyObject * dummy_custom_evaluator(PyThreadState *ts, _PyInterpreterFrame *frame, int throw)
> {
>     return _PyEval_EvalFrameDefault(ts, frame, throw);
> }
> ```
> 
> Thus for existing uses of the custom eval function, it might break existing implementations. Since this issue is also affecting, the `perf` support, I'm not sure if I should create another issue.

I think we should just use this issue and remove the perf-specific part

**Comment by art049** [comment]
Is this the expected behavior from the stack overflow protection put in place in #91079? (cc @markshannon)
I'd be up working on a fix regarding this but do you think the behavior should indeed be changed?

I managed a dirty workaround incrementing the `c_recursion_remaining` variable in the custom frame evaluator but it's probably not what we want to do globally.
